File: Питаленко_Максим/src/ptree.py
#!/usr/bin/env python3
# файл: src/ptree.py

import os
import sys

def get_process_info(pid):
    """Получает информацию о процессе из /proc/pid/status"""
    try:
        with open(f"/proc/{pid}/status", "r") as f:
            content = f.read()
        
        name = None
        ppid = None
        
        for line in content.splitlines():
            if line.startswith("Name:"):
                name = line.split()[1]
            elif line.startswith("PPid:"):
                ppid = int(line.split()[1])
            
            if name is not None and ppid is not None:
                break
        
        return name, ppid
    except (FileNotFoundError, PermissionError, ValueError):
        return None, None

def print_process_tree():
    """Печатает цепочку процессов от текущего до init"""
    current_pid = os.getpid()
    chain = []
    
    # Собираем информацию о текущем процессе
    name, _ = get_process_info(current_pid)
    if name:
        chain.append((name, current_pid))
    
    # Поднимаемся по цепочке родителей
    ppid = current_pid
    while ppid > 1:
        name, ppid = get_process_info(ppid)
        if name is None or ppid is None:
            break
        chain.append((name, ppid))
    
    # Форматируем вывод
    if chain:
        # Первый элемент - текущий процесс, остальные - родители
        result = []
        for i, (name, pid) in enumerate(chain):
            if i == 0:
                result.append(f"{name}({pid})")
            else:
                result.append(f" ← {name}({pid})")
        
        # Добавляем init/systemd в конец
        result.append(" ← systemd(1)")
        print("Process tree:", "".join(result))
    else:
        print("Could not determine process tree")

def main():
    print_process_tree()

if __name__ == "__main__":
    main()

File: lab2/gr9-sub1/RozhdaevEV/src/ptree.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define MAX_LINE 256

void print_process_tree(pid_t pid) {
    char path[64];
    char line[MAX_LINE];
    char name[64];
    pid_t ppid;
    FILE *file;
    
    while (pid > 1) {
 
        snprintf(path, sizeof(path), "/proc/%d/status", pid);
        file = fopen(path, "r");
        if (file == NULL) {
            printf(" ← [process %d not found]", pid);
            break;
        }
        
        ppid = -1;
        while (fgets(line, sizeof(line), file)) {
            if (strncmp(line, "Name:", 5) == 0) {
                sscanf(line, "Name:\t%s", name);
            } else if (strncmp(line, "PPid:", 5) == 0) {
                sscanf(line, "PPid:\t%d", &ppid);
            }
        }
        fclose(file);
        
        if (ppid == -1) {
            printf("%s(%d)", name, pid);
            break;
        } else {
            printf("%s(%d) ← ", name, pid);
        }
        
        pid = ppid;
    }
    printf("init/systemd(1)\n");
}

int main() {
    pid_t current_pid = getpid();
    printf("Process tree for PID %d:\n", current_pid);
    print_process_tree(current_pid);
    return 0;
}
